### PR TITLE
MLDB-1367 id encoding / utf-8 support

### PR DESCRIPTION
**Pull request by FinchPowers**

### CODE REVIEW

File: rest/rest_request_params.cc [hunk 1]
```diff
@@ -34,7 +35,14 @@ std::string restDecode(std::string str, std::string *)
...
 {
-    return std::move(str);
+    CURL *curl = curl_easy_init();
+    int outlength;
+    std::unique_ptr<char, void (*)(char *)> cres(
+        curl_easy_unescape(curl, str.rawData(), str.length(), &outlength),
+        [](char *p) { curl_free(p); });
+    Utf8String utf8(cres.get(), cres.get() + outlength);
+    curl_easy_cleanup(curl);
+    return std::move(utf8);
```
Comment (FinchPowers) [inline rest/rest_request_params.cc]: 1. Je suppose que je devrais me faire un curl_easy_init que je partage à travers ce fichier plutôt que de le réinstantier/détruire à chaque fois? Est-ce qu'il y en a déjà un disponible que je devrais utiliser?
2. Si je réutilise le même, est-ce que quelqu'un sait si c'est thread safe?
3. J'ai laissé le std::move mais de ce que je comprend il ne sert à rien.

Comment (jeremybarnes) [inline rest/rest_request_params.cc]: 1.  I would avoid using CURL if possible.
2.  I'm sure we're not looking at the right place here.  str is *already* a UTF-8 string, isn't it?  If we're sure it must be an ASCII encoded string, it should be a std::string; the fact it's already UTF-8 tells the user it's already been decoded.

1.  I think that curl_easy_init is already called elswhere; you should be able to grep it
2.  It's thread safe
3.  The std::move in the original version is useful, but not in your version, due to compiler optimizations for returning a value that's a local on the stack.

File: types/testing/decode_uri_test.cc [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+#endif
+
+#if TEST_ALL
+BOOST_AUTO_TEST_CASE(test_utf8_repetitions)
+{
+    Utf8String in = "%C3%A9%C3%A9";
+    Utf8String expected = "éé";
+    BOOST_CHECK_EQUAL(Url::decodeUri(in), expected);
+}
+#endif
```
Comment (FinchPowers) [inline types/testing/decode_uri_test.cc]: Si on passe du input invalide, par exmpele `%C3`, ça throw `Invalid sequence within utf-8 string`. Est-ce correct?

File: types/url.cc [hunk 3]
```diff
@@ -256,6 +257,73 @@ reconstitute(ML::DB::Store_Reader & store)
...
+            ++it; // over high
+            if (it == in.end() || !isxdigit(*it)) {
+                throw ML::Exception("Invalid uri"); // Todo better msg?
+            }
+            high = *it;
+            high -= high <= '9' ? '0' : (high <= 'F' ? 'A' : 'a') - 10;
+
+            ++it; // over low
+            if (it == in.end() || !isxdigit(*it)) {
+                throw ML::Exception("Invalid uri"); // Todo better msg?
```
Comment (FinchPowers) [inline types/url.cc]: Est-ce que `Invalid uri` est suffisant comme message d'erreur?

Comment (nicolaskruchten) [inline types/url.cc]: Badly encoded URI fragment: pata%%%ppp

File: types/testing/decode_uri_test.cc (hunk 2 above)
Comment (FinchPowers) [inline types/testing/decode_uri_test.cc]: Nouvelle erreur.
`what:   Invalid encoding on uri fragment: %C3`

**Comment by nicolaskruchten** [comment]
very good test coverage. +1

**Comment by nicolaskruchten** [comment]
but out of curiosity: were we not able to re-use the existing query-string decoding code? Was that just because of the issue with "+" ?

**Comment by FinchPowers** [comment]
I found no decoding code of any sort.

**Comment by nicolaskruchten** [comment]
that's odd, because we already supported URL encoding in query-strings... might be worth checking to see if these can be combined.

**Comment by nicolaskruchten** [comment]
i.e we could already send the query `select 'françois'` and it would decode and everything

File: types/url.cc [hunk 4]
```diff
@@ -256,6 +257,85 @@ reconstitute(ML::DB::Store_Reader & store)
 }
 #endif
 
+/**
+ * Decodes uri encoded with Percent-encoding. It is meant to act like
+ * JavaScript decodeURI.
+ **/
+Utf8String
+Url::
+decodeUri(Utf8String in)
```
Comment (guyd) [inline types/url.cc]: So we could not find a function that would do that?

Comment (FinchPowers) [inline types/url.cc]: En fouillant pour encoder les uri, j'ai trouvé je crois dans la librairie google url. Les coquins ne nomment pas leurs affaires camel case donc c'est plus dur à trouver avec des greps.

**Pull request merged**